core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java

```diff
@@ -151,6 +157,95 @@ protected void create(FateId fateId, FateKey fateKey) {
         + " and fateKey " + fateKey + " after " + maxAttempts + " attempts");
   }
 
+  @Override
+  public Optional<FateTxStore<T>> tryReserve(FateId fateId) {
+    // Create a unique FateReservation for this reservation attempt
+    FateReservation reservation = FateReservation.from(lockID, UUID.randomUUID());
+
+    FateMutator.Status status = newMutator(fateId).putReservedTx(reservation).tryMutate();
+    if (status.equals(FateMutator.Status.ACCEPTED)) {
+      return Optional.of(new FateTxStoreImpl(fateId, reservation));
+    } else if (status.equals(FateMutator.Status.UNKNOWN)) {
+      // If the status is UNKNOWN, this means an error occurred after the mutation was
+      // sent to the TabletServer, and it is unknown if the mutation was written. We
+      // need to check if the mutation was written and if it was written by this
+      // attempt at reservation. If it was written by this reservation attempt,
+      // we can return the FateTxStore since it was successfully reserved in this
+      // attempt, otherwise we return empty (was written by another reservation
+      // attempt or was not written at all).
+      try (Scanner scanner = context.createScanner(tableName, Authorizations.EMPTY)) {
+        scanner.setRange(getRow(fateId));
+        scanner.fetchColumn(TxColumnFamily.RESERVATION_COLUMN.getColumnFamily(),
+            TxColumnFamily.RESERVATION_COLUMN.getColumnQualifier());
+        FateReservation persistedRes = scanner.stream()
+            .filter(entry -> FateReservation.isFateReservation(entry.getValue().toString()))
+            .map(entry -> FateReservation.from(entry.getValue().toString())).findFirst()
```

Comment:
Looking into this a bit more, I think the reservation column may always exist and have a not reserved value.  So is the filter needed for that case?

Reply:
Yes, the reservation column always exists and the filter is needed for the case where it is not reserved. The value stored if the fate id is not reserved is FateMutatorImpl.NOT_RESERVED = "", and this is the value set when a new fate id is created and when a reserved fate id is unreserved. Thinking about this again, it would probably be better to just delete/create the column as needed instead of storing a not reserved value. What do you think?

Reply:
Yeah, I also think deleting would be better. I will make this change.

Reply:
>If going with the NOT_RESERVED value approach it would be good to make the validation of the column value more strict, will make a comment about that elsewhere.

Nevermind the code does not work exactly the way I thought it did. I was thinking when reading reservations could do something like the following

```java
if(value look like <lock id>:<uuid) {
  // process reserved
} else if (value is NOT_RESERVED) {
  //process not reserved
} else {
   throw new IllegalArgumentException("invalid reseved column value");
}
```

However the is not a place in the code that reads reservations in this way.  Was thinking the code should fail if the reserved column value was something like '`1234` because it does not look like a reservation and it also does not look like NOT_RESERVED

Reply:
If the reservation parsing code is not really strict and the code writing reservations has a bug, then could end with the case were reservation is silently ignored because it did not have the correct format.  Would be better to fail rather than ignore things that are not recognized.  Some of the code seems to have the following pattern

```java
f(value look like <lock id>:<uuid) {
  // process reserved
} else {
  // assume not reserved, but could have been an incorrectly written reservation
}
```

Reply:
Ah I see. Yeah I don't think I ever check if something is reserved like that and instead use a conditional mutation in FateMutatorImpl to reserve/unreserve. Should be pretty straightforward to change to create/delete the column on reserve/unreserve

Reply:
You're referring to MetaFateStore where I have that pattern correct? I don't think UserFateStore would have this issue but I could be wrong

Reply:
I'm not sure I follow how the reservation could be incorrectly written. I could just be missing/misunderstanding something but for MetaFateStore where I do this check, it checks if a FateReservation object is present and that can only take on the form <lock id>:<uuid>. Creating the FateReservation object handles these checks of making sure only a valid FateReservation of the form <lock id>:<uuid> is created.

Reply:
Made this comment as an example

https://github.com/apache/accumulo/pull/4524/files?diff=unified&w=1#r1672972761

I am not sure if MetaFateStore has any problems, was only thinking through the UserFateStore code.  Ideally both should be very strict w/ reservation parsing.

Reply:
Ah I understand now. Yes that should (but currently doesn't) throw an error if the entry is not NOT_RESERVED and not a valid FateReservation str. Since I will be changing the column to only ever include a valid FateReservation (column created and deleted as needed), I will just change this to throw an error if it sees anything but a FateReservation string. Thanks for the clarification

Reply:
I had made a comment elsewhere about avoiding using `toString()` for serialization and suggested using `getSerialized()` instead  When I made that comment I thought the impl of `toString()` and `getSerialized()` were the same.  Now I realize they are different.  Maybe part of the solution to this is converging UserFateStore and MetaFateStore on the same serialization impl and making the de-serialization as strict as possible. Would that work?  If NOT_RESERVED is dropped seems like that makes it easier for them to have same ser/deser impl.

Reply:
Yeah I think that would work and sounds like a good idea. It would clean up/get rid of some code. Don't think converting the FateReservation to/from a string would be needed anymore.

Reply:
I attempted to refactor UserFateStore to create/delete the RESERVATION column on reserve/unreserve, however, I ran into an issue that I could not figure out, so I ended up reverting this change. Multiple threads would occassionally reserve the same fate id. I could not figure out how/why this was occuring. I think the switch to deleting the column should be made, but maybe as a follow on or after discussion about the potential cause of this. Maybe I could put up a new commit with just this change to easily see the difference and what the issue might be.

Reply:
> Multiple threads would occassionally reserve the same fate id.

Not sure what the problem was, but its kinda neat you detected that bug.  Where you running an IT when you saw multiple threads reserve a transaction?

Reply:
Yes, I was running my new `MultipleStoresIT.testDeadReservationsCleanup()` test, which would occasionally fail with my new changes. The test works as follows:

1) We have a fate object fate1 which is used to have 4 (the default MANAGER_FATE_THREADPOOL_SIZE) workers reserve 4 transactions with lock1
2) We then simulate that the imaginary manager dies and a new one starts, so lock1 is no longer held instead lock2 is now held
3) A new fate2 is created which spins up 4 more workers which should eventually reserve the 4 transactions so long as the DeadReservationCleaner thread unreserves the transactions reserved with lock1

Took a while to track down, but found that when waiting for the transactions to be reserved with lock2, multiple threads would sometimes reserve the same transaction, so not all 4 transactions would be reserved causing the test to fail. Found that the issue was due to the change from storing an unreserved value in the RESERVATION column to just deleting/creating the column on unreserve/reserve.

I could not figure out why or how this could occur with these changes. I could potentially put a new commit up at some point with these changes to see if you could find anything wrong with how I am doing it.

Reply:
This conversion is really long, but it seems like everything in it is resolved.  @kevinrr888 can you resolve this if you also don't see anything outstanding in the conversations?

Reply:
Yes, resolved. This conversation will be linked in the new issue to address this change as context.